internal/planner/mapper/mapper.go

```diff
@@ -1296,87 +1296,95 @@ func ToFilter(source request.Filter, mapping *core.DocumentMapping) *Filter {
-			Index: index,
+		key = &PropertyIndex{
+			Index: indexes[0],
+		}
+	} else {
+		// Key is an object property
+		key = &ObjectProperty{
+			Name: sourceKey,
 		}
-		switch typedClause := sourceClause.(type) {
+	}
+
+	switch typedClause := sourceClause.(type) {
+	case []any:
+		return key, toFilterList(typedClause, mapping)
+
+	case map[string]any:
+		return key, toFilterMap(key, typedClause, mapping)
+
+	default:
+		return key, typedClause
+	}
+}
+
+func toFilterMap(
+	sourceKey connor.FilterKey,
+	sourceClause map[string]any,
+	mapping *core.DocumentMapping,
+) map[connor.FilterKey]any {
+	prop, isProp := sourceKey.(*PropertyIndex)
+	innerMapClause := make(map[connor.FilterKey]any)
+	for innerSourceKey, innerSourceValue := range sourceClause {
+		var innerMapping *core.DocumentMapping
+		// innerSourceValue may refer to a child mapping or
+		// an inline array if we don't have a child mapping
+		_, ok := innerSourceValue.(map[string]any)
+		if ok && isProp && prop.Index < len(mapping.ChildMappings) {
+			// If the innerSourceValue is also a map, then we should parse the nested clause
+			// using the child mapping, as this key must refer to a host property in a join
+			// and deeper keys must refer to properties on the child items.
+			innerMapping = mapping.ChildMappings[prop.Index]
+		} else {
+			innerMapping = mapping
+		}
+		rKey, rValue := toFilterKeyValue(innerSourceKey, innerSourceValue, innerMapping)
+		innerMapClause[rKey] = rValue
+	}
+	return innerMapClause
+}
+
+func toFilterList(sourceClause []any, mapping *core.DocumentMapping) []any {
+	returnClauses := make([]any, len(sourceClause))
+	for i, innerSourceClause := range sourceClause {
+		switch typedInnerSourceClause := innerSourceClause.(type) {
 		case map[string]any:
-			returnClause := map[connor.FilterKey]any{}
-			for innerSourceKey, innerSourceValue := range typedClause {
-				var innerMapping *core.DocumentMapping
-				// innerSourceValue may refer to a child mapping or
-				// an inline array if we don't have a child mapping
-				_, ok := innerSourceValue.(map[string]any)
-				if ok && index < len(mapping.ChildMappings) {
-					// If the innerSourceValue is also a map, then we should parse the nested clause
-					// using the child mapping, as this key must refer to a host property in a join
-					// and deeper keys must refer to properties on the child items.
-					innerMapping = mapping.ChildMappings[index]
-				} else {
-					innerMapping = mapping
-				}
-				rKey, rValue := toFilterMap(innerSourceKey, innerSourceValue, innerMapping)
-				returnClause[rKey] = rValue
+			innerMapClause := map[connor.FilterKey]any{}
+			for innerSourceKey, innerSourceValue := range typedInnerSourceClause {
+				rKey, rValue := toFilterKeyValue(innerSourceKey, innerSourceValue, mapping)
+				innerMapClause[rKey] = rValue
 			}
-			return key, returnClause
+			returnClauses[i] = innerMapClause
 		default:
-			return key, sourceClause
+			returnClauses[i] = innerSourceClause
```

Comment:
I think this statement is unreachable. Only compound operators (`_and`, `_not`, `_or`) will call this function and their GQL types will already be validated before reaching the mapper.

Would you prefer for the type switch to be removed?

Reply:
I'm referring to this line in general. It's called for all compound filter operations that contain an array, but this line in particular should never be called because the filter would have failed GQL validation.

Reply:
Ah okay, I think I understand - in that case I'm probably in favour of removing the line